src/main/java/spoon/support/adaption/TypeAdaptor.java

```diff
@@ -0,0 +1,598 @@
+import spoon.reflect.declaration.CtParameter;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.CtTypeParameter;
+import spoon.reflect.reference.CtArrayTypeReference;
+import spoon.reflect.reference.CtTypeParameterReference;
+import spoon.reflect.reference.CtTypeReference;
+import spoon.support.visitor.ClassTypingContext;
+import spoon.support.visitor.MethodTypingContext;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Determines subtyping relationships and adapts generics from a super- to a subclass.
+ */
+public class TypeAdaptor {
+
+	private final CtType<?> hierarchyStart;
+	private final CtTypeReference<?> hierarchyStartReference;
+	private final boolean initializedWithReference;
+	private CtMethod<?> startMethod;
+	private ClassTypingContext oldClassTypingContext;
+
+	/**
+	 * Creates a new type adaptor using the given type as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtType<?> hierarchyStart) {
+		this.hierarchyStart = hierarchyStart;
+		this.hierarchyStartReference = hierarchyStart.getReference();
+		this.initializedWithReference = false;
+	}
+
+	/**
+	 * Creates a new type adaptor using the given reference as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtTypeReference<?> hierarchyStart) {
+		CtTypeReference<?> usedHierarchyStart = hierarchyStart;
+		if (hierarchyStart instanceof CtArrayTypeReference) {
+			usedHierarchyStart = ((CtArrayTypeReference<?>) hierarchyStart).getArrayType();
+		}
+		this.hierarchyStartReference = usedHierarchyStart;
+		this.hierarchyStart = hierarchyStartReference.getTypeDeclaration();
+		this.initializedWithReference = true;
+	}
+
+	/**
+	 * Creates a new type adaptor using the given method as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtMethod<?> hierarchyStart) {
+		this(hierarchyStart.getDeclaringType());
+		this.startMethod = hierarchyStart;
+	}
+
+	/**
+	 * Checks if the context of this type adapter is a subtype of the passed superRef.
+	 *
+	 * @param superRef the super reference to check against
+	 * @return true if the context of this type adapter is a subtype of the passed superRef
+	 * @implNote This implementation behaves the same as {@code isSubtype(hierarchyStart,
+	 * 	superRef)}
+	 * @see #isSubtype(CtType, CtTypeReference)
+	 */
+	@SuppressWarnings("removal")
+	public boolean isSubtypeOf(CtTypeReference<?> superRef) {
+		if (superRef.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return getOldClassTypingContext().isSubtypeOf(superRef);
+		}
+		// We have no declaration so we can't really do any subtype queries
+		if (hierarchyStart == null) {
+			return false;
+		}
```

Comment:
Is it wise for `null` to be allowed for `hierarchyStart`? It seems like nonsense.

Reply:
The context can be initialized from a reference to a type not on the classpath. In that case we don't have a corresponding declaration. The current implementation just rejects such queries as they are rather pointless. I will amend the comment, do you have a better idea for handling this situation? We could store an optional but I am not sure whether that's strictly better.